clustering/infinispan/spi/src/main/java/org/wildfly/clustering/infinispan/spi/affinity/DefaultKeyAffinityService.java

```diff
@@ -0,0 +1,301 @@
+        this.cache.addListener(this);
+    }
+
+    @Override
+    public void stop() {
+        this.cache.removeListener(this);
+        WildFlySecurityManager.doUnchecked(this.executor, DefaultExecutorService.SHUTDOWN_NOW_ACTION);
+    }
+
+    @Override
+    public K getCollocatedKey(K otherKey) {
+        KeyAffinityState<K> currentState = this.currentState.get();
+        if (currentState == null) {
+            // Not yet started!
+            throw new IllegalStateException();
+        }
+        return this.getCollocatedKey(currentState, otherKey);
+    }
+
+    private K getCollocatedKey(KeyAffinityState<K> state, K otherKey) {
+        K key = this.poll(state.getRegistry(), state.getDistribution().getPrimaryOwner(otherKey));
+        if (key != null) {
+            return key;
+        }
+        KeyAffinityState<K> currentState = this.currentState.get();
+        // If state is out-dated, retry
+        if (state != currentState) {
+            return this.getCollocatedKey(currentState, otherKey);
+        }
+        LOGGER.debugf("Could not obtain pre-generated key with same affinity as %s -- generating random key", otherKey);
+        return this.generator.getKey();
+    }
+
+    @Override
+    public K getKeyForAddress(Address address) {
+        if (!this.filter.test(address)) {
+            throw new IllegalArgumentException(address.toString());
+        }
+        KeyAffinityState<K> currentState = this.currentState.get();
+        if (currentState == null) {
+            // Not yet started!
+            throw new IllegalStateException();
+        }
+        return this.getKeyForAddress(currentState, address);
+    }
+
+    private K getKeyForAddress(KeyAffinityState<K> state, Address address) {
+        K key = this.poll(state.getRegistry(), address);
+        if (key != null) {
+            return key;
+        }
+        KeyAffinityState<K> currentState = this.currentState.get();
+        // If state is out-dated, retry
+        if (state != currentState) {
+            return this.getKeyForAddress(currentState, address);
+        }
+        LOGGER.debugf("Could not obtain pre-generated key with affinity for %s -- generating random key", address);
+        return this.generator.getKey();
+    }
+
+    private K poll(KeyRegistry<K> registry, Address address) {
+        BlockingQueue<K> keys = registry.getKeys(address);
+        if (keys != null) {
+            Duration timeout = this.timeout;
+            long nanos = (timeout.getSeconds() == 0) ? timeout.getNano() : timeout.toNanos();
+            try {
+                return keys.poll(nanos, TimeUnit.NANOSECONDS);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+        return null;
+    }
+
+    @TopologyChanged
+    public CompletionStage<Void> viewChanged(TopologyChangedEvent<?, ?> event) {
```

Comment:
This method should rather be called `topologyChanged(...)`.

Reply:
The name is arbitrary, but I agree with you.